### PR TITLE
Add support for AutomationPeer

### PR DESCRIPTION
**Pull request by miguelrochefort**

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Automation/Peers/AutomationPeer.cs [hunk 1]
```diff
@@ -1,15 +1,165 @@
...
+using Windows.UI.Xaml.Automation.Provider;
+
 namespace Windows.UI.Xaml.Automation.Peers
 {
-	public  partial class AutomationPeer : global::Windows.UI.Xaml.DependencyObject
+	public partial class AutomationPeer : DependencyObject
 	{
+		#region Public
+
+		public void InvalidatePeer()
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/Automation/Peers/AutomationPeer.cs]: [NotImplemented]?

File: src/Uno.UI/UI/Xaml/Automation/Peers/ComboBoxAutomationPeer.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
+using Uno;
+using Windows.UI.Xaml.Controls;
+
+namespace Windows.UI.Xaml.Automation.Peers
+{
+	public  partial class ComboBoxAutomationPeer : SelectorAutomationPeer, Provider.IExpandCollapseProvider, Provider.IValueProvider, Provider.IWindowProvider
+	{
+		public ExpandCollapseState ExpandCollapseState
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/Automation/Peers/ComboBoxAutomationPeer.cs]: [NotImplemented]

File: src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.Android.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		}
+
+		private bool GetScrollable()
+		{
+			return this is IScrollProvider;
+		}
+
+		private int GetMaxTextLength()
+		{
+			return Owner is TextBox textBox ? textBox.MaxLength : 0;
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.Android.cs]: Why is this not virtual? Should there be a TODO to use ITextProvider?

File: src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.cs [hunk 4]
```diff
@@ -1,34 +1,147 @@
...
+			if (GetLabeledBy() is AutomationPeer labelAutomationPeer && labelAutomationPeer.GetName() is string label && !string.IsNullOrEmpty(label))
+			{
+				return label;
+			}
+
+			if (FeatureConfiguration.AutomationPeer.UseSimpleAccessibility)
+			{
+				if (Owner is View view && AutomationProperties.GetAccessibilityView(Owner) != AccessibilityView.Raw)
+				{
+					// TODO: How expensive is this? Is this called recursively for all children?
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.cs]: Is it?

File: src/Uno.UI/UI/Xaml/Controls/ContentControl/ImplicitTextBlock.cs [hunk 5]
```diff
@@ -15,7 +15,11 @@ namespace Windows.UI.Xaml.Controls
 	[Data.Bindable]
 	public partial class ImplicitTextBlock : TextBlock
 	{
-		public ImplicitTextBlock() { }
+		public ImplicitTextBlock(DependencyObject parent)
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/Controls/ContentControl/ImplicitTextBlock.cs]: Is the reference to the ContentControl being held? Ie is there any risk of a memory leak?

File: src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs [hunk 6]
```diff
@@ -70,5 +64,27 @@ public void OnTemplateRecycled()
 		{
 			IsChecked = false;
 		}
+
+		protected virtual void OnToggle()
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs]: OnAutomationToggle()? Is it used/useful in any other context?

File: src/Uno.UI/UI/Xaml/IFrameworkElement.cs [hunk 7]
```diff
@@ -152,10 +157,12 @@ public static void Initialize(IFrameworkElement e)
...
 			}
 
-#if XAMARIN_ANDROID
-			if (Debugger.IsAttached)
+#if __ANDROID__
+			if (e is View view)
 			{
-				(e as View).ContentDescription = e.GetType().ToString();
+				// TODO: This causes Android to cycle through every visible IFrameworkElement every time the accessibility focus is changed
+				// and calls the overriden InitializeAccessibilityNodeInfo method. This could eventually cause performance issues when accessibility is enabled.
```
Comment (Dissolute) [inline src/Uno.UI/UI/Xaml/IFrameworkElement.cs]: When is 'eventually'? Is this a serious performance concern?

File: doc/articles/working-with-accessibility.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+## Tips
+
+- Always set and localize `AppBarButton.Label` (even if it's not displayed on Android and iOS). It is used by the screen reader for accessibility.
+- Always localize `AutomationProperties.Name`. The name of the resource should look like this: `MyButton.[using:Windows.UI.Xaml.Automation]AutomationProperties.Name`.
+- Avoid using `Opacity="0"` and `IsHitTestVisible="False"` when you can use `Visibility="Collapsed"`. The screen reader can still focus the former, but not the latter.
+- Avoid stacking `TextBlock`s inside  a `Panel` when you can use `Inlines` inside a `TextBlock` (using `LineBreak` if necessary). This allows the screen reader to read all the text at once, instead of having the user select every part manually.
+- Use a converter to trim long text. While a `TextBlock` might ellipsize long text, the screen reader will read the entire text provided.
+- Avoid creating custom controls when you can use built-in ones. If you must, make sure to implement and provide an appropriate `AutomationPeer`.
+- You can disable accessibility focus of native elements using `android:ImportantForAccessibility="No"` and `ios:IsAccessibilityElement="False"`.
```
[the remark this answers is not in the document]
Comment (miguelrochefort) [reply doc/articles/working-with-accessibility.md]: You can disable accessibility focus of UWP/Uno elements using `AutomationProperties.AccessibilityView="Raw"`. 

It won't work on native elements, hence the `ImportantFocAccessibility`/`IsAccessibilityElement` workaround.

File: src/Uno.UI/UI/Xaml/Controls/ContentControl/ImplicitTextBlock.cs (hunk 5 above)
Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/Controls/ContentControl/ImplicitTextBlock.cs]: No. It only needs its parent to retrieve its `AutomationProperties.AccessibilityView` value once.

File: src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs (hunk 6 above)
Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/Controls/Primitives/ToggleButton.cs]: `OnToggle` is part of `ToggleButton`'s contract.

`AutomationPeerToggle` exists to allow `OnToggle` to be called from automation peers.

File: src/Uno.UI/UI/Xaml/IFrameworkElement.cs (hunk 7 above)
Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/IFrameworkElement.cs]: 'eventually' is if we have complex UI with lots of elements. I've only noticed lags when cycling through accessible elements in debug.

The performance problem seem to be due to interop. It should only affect focus change on devices with TalkBack turned on. A possible optimization could be to set this value to false if an element's `CreateAutomationPeer` returns `null`.

File: src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.Android.cs (hunk 3 above)
Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.Android.cs]: I don't know if there's an equivalent in `AutomationPeer`'s API.

Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.Android.cs]: I can't find anything about maximum text length at first glance.

File: doc/articles/working-with-accessibility.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+- [Android] The accessible name of the native back button can't be customized.
+- [Android] `TextBox` and `PasswordBox` don't hint "double tap to edit".
+- [Android] You can tap to focus an accessible child of an accessible element that has never been focused even if `UseSimpleAccessibility` is enabled.
+- `AutomationPeer.GetLocalizedControlType()` is not localized.
+- `FlipViewItem` reads "double tap to activate" even if it's not interactive.
+- Navigating through accessible elements can slow down when the UI is complex.
+- You can't cycle through `FlipView` items using accessibility focus.
+- There is no way to control or predict accessibility focus following a navigation (page or modal).
+- `Control`s don't announce state changes.
+- Only the following elements support accessibility:
```
Comment (carldebilly) [inline doc/articles/working-with-accessibility.md]: supports

File: src/Uno.UI/UI/Xaml/Automation/Peers/AutomationControlTypes.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
+﻿namespace Windows.UI.Xaml.Automation.Peers
+{
+	public enum AutomationControlType
```
[the remark this answers is not in the document]
Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/Automation/Peers/AutomationControlTypes.cs]: The order is important. I added the values.

File: src/Uno.UI/Extensions/UIViewExtensions.iOS.cs [hunk 11]
```diff
@@ -197,7 +197,20 @@ public static IEnumerable<UIView> FindSubviews(this UIView view, Func<UIView, bo
 		/// <returns>A lazy enumerable of views</returns>
 		public static IEnumerable<UIView> EnumerateAllChildren(this UIView view, int maxDepth = 20)
 		{
-			return FindSubviews(view, _ => true, maxDepth);
+			var children = view.GetChildren().OfType<UIElement>();
```
[the remark this answers is not in the document]
Comment (miguelrochefort) [reply src/Uno.UI/Extensions/UIViewExtensions.iOS.cs]: Fixed.

File: src/SourceGenerators/Uno.UI.Tasks/ResourcesGenerator/AndroidResourcesWriter.cs [hunk 12]
```diff
@@ -17,7 +17,7 @@ public static void Write(Dictionary<string, string> resources, string path, stri
 				new XElement("resources", resources.Select(resource =>
 					new XElement("string",
 						new XAttribute("formatted", "false"), // allows special characters (%, $, etc.)
-						new XAttribute("name", resource.Key),
+						new XAttribute("name", AndroidResourceNameEncoder.Encode(resource.Key)),
```
[the remark this answers is not in the document]
Comment (miguelrochefort) [reply src/SourceGenerators/Uno.UI.Tasks/ResourcesGenerator/AndroidResourcesWriter.cs]: I created a different PR: #156

File: src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.cs (hunk 4 above)
Comment (miguelrochefort) [reply src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.cs]: I refactored that part, confirmed this wasn't a problem, and explained why.

**Pull request merged**